### PR TITLE
Add CRD tooling and a Kafka CRD

### LINKED ISSUE #490: Kafka CRD Schema + Java Marshalling
We need to decide on the schema for a Kafka CRD. This includes:

* the validation schema used in the `CustomResourceDefinition`
* the marshalling on instance resources to and from a POJO representation
* tests for the above

It does not include refactoring the operator to actually use the POJO representation.

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Enhancement / new feature

### Description

This PR fixes #490 and #497.

This PR:

* Adds a tool for generating a `CustomResourceDefinition` from annotated POJOs. The tool uses Java reflection to do this and understands Jackson annotations and adds a few of its own.
* Adds a tool for generating user documentation of the CRD from the same annotated POJOs. The output is in asciidoc format and is broadly modelled on the K8S API reference in how the documentation is structured. 
* Moves some of the existing model classes (those which are represented as JSON in the cluster CM) into a different package
* Adds some new classes (e.g. `Kafka`, `Zookeeper`) to that same package in order to arrive at the POJOs for a `Kafka` CRD (represented by the `KafkaAssembly` class).
* Applies the two tools to this `KafkaAssembly`
* Adds tests that we can round-trip from a tree of the POJOs to YAML, and an integration test that we can create the CRD in K8S and then create an instance document.

We will need better naming as the new package and classes have created a `KafkaAssembly` class (representing the `Kafka` CRD), a `Kafka` class (representing the Kafka part of the CRD). The existing `KafkaCluster` is a model built on top of this representation. I also don't especially like the name for the new package (the CRD is really an implementation detail). 

So I wonder if something like `io.strimzi.api.kafka` would make sense (with `io.strimzi.api.kafkaconnect`, `io.strimzi.api.kafkaconnect.s2i` and `io.strimzi.api.topic` to come) for the new package. And rename `KafkaCluster` → `KafkaResources` (since really that class is about determining the K8S resources reflecting the K8S resources as represented by the CRD/API).

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/KafkaAssembly.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                scope = "Namespaced",
+                version = "v1alpha1"
+        )
+)
+@JsonPropertyOrder({"apiVersion", "kind", "metadata", "spec"})
+public class KafkaAssembly extends CustomResource {
+
+    private static final long serialVersionUID = 1L;
+    public static final String RESOURCE_KIND = "Kafka";
+    public static final String RESOURCE_GROUP = "cluster-operator.strimzi.io";
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/KafkaAssembly.java]: What about calling this just `cluster.strimzi.io`? The `operator` seem a bit unnecessary.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/KafkaAssembly.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                group = KafkaAssembly.RESOURCE_GROUP,
+                scope = "Namespaced",
+                version = "v1alpha1"
+        )
+)
+@JsonPropertyOrder({"apiVersion", "kind", "metadata", "spec"})
+public class KafkaAssembly extends CustomResource {
+
+    private static final long serialVersionUID = 1L;
+    public static final String RESOURCE_KIND = "Kafka";
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/KafkaAssembly.java]: Should this really start with capital letter?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConfiguration.java [hunk 3]
```diff
@@ -5,37 +5,24 @@
...
+import java.util.Map;
 
 import static java.util.Arrays.asList;
 
 /**
  * Class for handling Kafka configuration passed by the user
  */
 public class KafkaConfiguration extends AbstractConfiguration {
+
+    public static final String FORBIDDEN_PREFIXES = "listeners, advertised., broker., listener., host.name, port, "
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConfiguration.java]: Why this change? It seems to make it quite hard to read. And why just here and not in `KafkaConnectConfioguration`?

File: crd-generator/src/main/java/io/strimzi/crdgenerator/OpenShiftOriginLinker.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+    private final String baseUrl;
+
+    public OpenShiftOriginLinker(String baseUrl) {
+        //https://docs.openshift.org/3.9/rest_api/
+        this.baseUrl = baseUrl;
+    }
+    @Override
+    public String link(KubeLink kubeLink) {
+        https://docs.openshift.org/3.9/rest_api/apis-extensions/v1beta1.DaemonSet.html
```
Comment (scholzj) [inline crd-generator/src/main/java/io/strimzi/crdgenerator/OpenShiftOriginLinker.java]: Shouldn't this be a comment?

File: examples/install/crd/kafka-instance.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+apiVersion: cluster-operator.strimzi.io/v1alpha1
+kind: Kafka
+metadata:
+  name: example
+spec:
+  topicOperator:
+    - foo: sfv
```
Comment (scholzj) [inline examples/install/crd/kafka-instance.yaml]: This seems to be a bit incomplete?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/AbstractSsLike.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            "See https://github.com/prometheus/jmx_exporter for details of the structure of this configuration.")
+    public Map<String, Object> getMetrics() {
+        return metrics;
+    }
+
+    public void setMetrics(Map<String, Object> metrics) {
+        this.metrics = metrics;
+    }
+
+    @Description("Persistent storage (disk). Cannot be updated.")
```
[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/AbstractSsLike.java]: I meant persistent in the sense of it not being RAM, basically. "Storage configuration (disk)." is a fairly useless description though: It's _obviously_ storage configuration, because the property is called "storage" and it's in an object which is about configuring a Kafka or Zookeeper cluster. 

Although `Storage` is a good name for the class perhaps the name of the property should be more along the line `logStorage`, since that's what we put there?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/AbstractSsLike.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    public Storage getStorage() {
+        return storage;
+    }
+
+    public void setStorage(Storage storage) {
+        this.storage = storage;
+    }
+
+    @Description("Pod affinity rules.")
+    @KubeLink(group = "core", version = "v1", kind = "affinity")
```
[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/AbstractSsLike.java]: > do we need to be more specific here? Or is this really enough?

That, precisely, do you mean by 'this'? The `@KubeLink` or the `@Description`?

I used `@KubeLink` here because I'm reusing the Fabric8 (and hence K8S) `affinity`, so it's precisely the same. That's not the case with resources because we currently have our own classes for that (because we only support "memory" and "cpu", whereas in principle K8S can support other things too (i.e it's a map in k8s)). Of course we might want to revisit that decision and reuse the Fabric8 resources, but not in this PR.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/KafkaAssembly.java (hunk 1 above)
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/KafkaAssembly.java]: I think we should put this whole API in a separate maven module, along with the other CRDs when we get them. Imho, the package name should be something like `io.strimzi.api.kafka`. Then the k8s group would be something like `kafka.strimzi.io`. I'm not sure that having the word operator in there is completely necessary.

**Comment by tombentley** [comment]
Thanks for the review @scholzj.

> It kind of seems to deliver bunch of dead code. What are the advantages of merging this?

It seems we've misunderstood each other. I had assumed it would be OK to open a PR for each part in order for the reviews to be manageable (while still being useful/meaningful). I can obviously put together a single PR for #490, #493, #497 and #498 if that's what you prefer, but personally I don't think such a PR could be reviewed in enough detail.

How about if we have a running feature branch and I open PRs for additions to that and we only merge it into master at the end. That way we can have incremental (+more manageable, digestible ) reviews for each PR into the feature branch and only at the end when we have a coherent feature do we merge it into master.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/AbstractSsLike.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/AbstractSsLike.java]: There is a lot of people who do not understand that the logs are also the records and not only the Log4j logs. So I think that would be even more confusing. To me it seems that _storage_ is pretty clear - just the _Persistent_ is not 100% correct.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/AbstractSsLike.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/AbstractSsLike.java]: With _this_ I meant the `@KubeLink`. I'm a bit wondering if this specification is enough given that fields are added to the Kubernetes API without bumping the versions.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/KafkaAssembly.java (hunk 1 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/KafkaAssembly.java]: `kafka.strimzi.io` sounds also good to me. Why do we need to wait for some changes to the project structure with this? We should ideally not change it after we introduce the CRDs, so we should change it right now.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/KafkaAssembly.java (hunk 2 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/crd/model/KafkaAssembly.java]: I guess you are right. I guess Kubernetes might be anyway treating it as case insensitive.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConfiguration.java (hunk 3 above)
Comment (scholzj) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConfiguration.java]: That sounds reasonable. Thanks.

**Comment by scholzj** [comment]
Well ... I don't think this breaks anything. It just doesn't do anything right now :-). So I'm flexible if we want to merge this. 

That said, I'm not sure there is really such a difference between two PRs each with 100 files or one PR with 200 files anymore:-). We should also try not to make a habit out of it.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    public static final String RESOURCE_LIST_KIND = RESOURCE_KIND + "List";
+    public static final String RESOURCE_GROUP = "kafka.strimzi.io";
+    public static final String RESOURCE_PLURAL = "kafkas";
+    public static final String RESOURCE_SINGULAR = "kafka";
+    public static final String RESOURCE_NAME = RESOURCE_PLURAL + "." + RESOURCE_GROUP;
+    public static final String CRD_API_VERSION = "apiextensions.k8s.io/v1beta1";
+    public static final String CRD_NAME = RESOURCE_PLURAL + "." + RESOURCE_GROUP;
+
+    private String apiVersion;
+    private ObjectMeta metadata;
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java]: It seems we are not using this field even because it's in the `CustomResource` base class and we are using getter/setter for that.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    private transient KafkaAssemblySpec spec;
+    private Map<String, Object> additionalProperties = new HashMap<>(0);
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java]: What's the purpose of this `additionalProperties` ?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java [hunk 10]
```diff
@@ -167,12 +174,40 @@ public static KafkaCluster fromConfigMap(ConfigMap kafkaClusterCm) {
...
             kafka.setRackConfig(rackConfig);
         }
-        kafka.setInitImage(Utils.getNonEmptyString(data, KEY_INIT_IMAGE, DEFAULT_INIT_IMAGE));
+        kafka.setInitImage(Utils.getNonEmptyString(data, KEY_INIT_IMAGE, Kafka.DEFAULT_INIT_IMAGE));
         kafka.setUserAffinity(Utils.getAffinity(data.get(KEY_AFFINITY)));
 
         return kafka;
     }
 
+    public static KafkaCluster fromCrd(KafkaAssembly kafkaAssembly) {
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: Javadoc

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/TopicOperator.java [hunk 11]
```diff
@@ -195,6 +192,29 @@ public static TopicOperator fromConfigMap(ConfigMap kafkaClusterCm) {
         return topicOperator;
     }
 
+
+    public static TopicOperator fromCrd(KafkaAssembly kafkaAssembly) {
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/TopicOperator.java]: Javadoc

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ZookeeperCluster.java [hunk 12]
```diff
@@ -144,6 +152,32 @@ public static ZookeeperCluster fromConfigMap(ConfigMap kafkaClusterCm) {
         return zk;
     }
 
+    public static ZookeeperCluster fromCrd(KafkaAssembly kafkaAssembly) {
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ZookeeperCluster.java]: Javadoc

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Storage.java [hunk 13]
```diff
@@ -15,15 +15,17 @@
...
+ * @deprecated Replaced with {@link io.strimzi.api.kafka.model.Storage}
  */
+@Deprecated
 public class Storage {
 
     public static final String TYPE_FIELD = "type";
     public static final String SIZE_FIELD = "size";
     public static final String STORAGE_CLASS_FIELD = "class";
     public static final String SELECTOR_FIELD = "selector";
     public static final String SELECTOR_MATCH_LABELS_FIELD = "match-labels";
```
[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/Storage.java]: The way it worked out I wrote a different, Jackson-annotated, `Storage` class which has a `matchLabels` property. This `Storage` class is now deprecated and I should be able to remove it (in this PR) once I've updated all the testing tests to use the `KafkaAssembly` rather than `ConfigMap`.

File: api/src/main/java/io/strimzi/api/kafka/model/AbstractSsLike.java [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+import java.util.Map;
+
+/**
+ * A baseclass for {@link Kafka} and {@link Zookeeper} with all their common properties.
+ */
+@Buildable(editableEnabled = false,
+        generateBuilderPackage = true,
+        builderPackage = "io.strimzi.api.kafka.model"
+)
+public abstract class AbstractSsLike {
```
[the remark this answers is not in the document]
Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/AbstractSsLike.java]: In the work I'd doing for the `KafkaConnect` kind I've reused this class again, and since the connect support isn't SS based I renamed it to `ReplicatedJvmPods`. It's a bit of a weak abstraction, to be honest, just serving to group together some properties, and saving us from having the declare them on the (now 3) different subclasses.

File: api/src/main/java/io/strimzi/api/kafka/model/AbstractSsLike.java [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+    protected Probe readinessProbe;
+
+    protected JvmOptions jvmOptions;
+
+    protected Map<String, Object> metrics = new HashMap<>(0);
+
+    protected Storage storage;
+
+    protected Affinity affinity;
```
[the remark this answers is not in the document]
Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/AbstractSsLike.java]: Yes, we do!

File: api/src/main/java/io/strimzi/api/kafka/model/Probe.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    @JsonAnyGetter
+    public Map<String, Object> getAdditionalProperties() {
+        return this.additionalProperties;
+    }
+
+    @JsonAnySetter
+    public void setAdditionalProperty(String name, Object value) {
+        this.additionalProperties.put(name, value);
+    }
+}
```
[the remark this answers is not in the document]
Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/Probe.java]: The fabric8 `Probe` is the full thing, with properties for `httpGet`, `exec`, `tpcSocket` as well as the delay, timeout. If we reused it people could mistakenly set those things and think it would make a difference. Really the CO is in charge of what the probe actually does, and we just want to give the user some control over how it gets run. Maybe for this reason the name Probe is confusing and we should name it `Healthcheck`, or something else.

File: api/src/main/java/io/strimzi/api/kafka/model/AbstractSsLike.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/AbstractSsLike.java]: So why not just `AbstractAssembly` ? Why the `Jvm' in the name ? Replicated pods sounds really weird to me.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+    private static final long serialVersionUID = 1L;
+    public static final String VERSION = "v1alpha1";
+    public static final String RESOURCE_KIND = "Kafka";
+    public static final String RESOURCE_LIST_KIND = RESOURCE_KIND + "List";
+    public static final String RESOURCE_GROUP = "kafka.strimzi.io";
+    public static final String RESOURCE_PLURAL = "kafkas";
+    public static final String RESOURCE_SINGULAR = "kafka";
+    public static final String RESOURCE_NAME = RESOURCE_PLURAL + "." + RESOURCE_GROUP;
+    public static final String CRD_API_VERSION = "apiextensions.k8s.io/v1beta1";
+    public static final String CRD_NAME = RESOURCE_PLURAL + "." + RESOURCE_GROUP;
```
[the remark this answers is not in the document]
Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java]: No. K8s imposes a naming scheme on what the CRD's `metadata.name` has to be.

File: api/src/main/java/io/strimzi/api/kafka/model/Probe.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/Probe.java]: Yes at this point. I would prefer a different name as `Healthcheck`.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java (hunk 8 above)
Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java]: This is actually working around an issue with sundrio, which is reported here: https://github.com/sundrio/sundrio/issues/104, but I hadn't realised until yesterday that overriding getMetadata() isn't enough, I actually have to have this field present too for it to generate a `withMetadata()` in the `KafkaAssemblyBuilder`. The issue is really about getting it to generate a `withNewMetadata()` method too.

File: api/src/main/java/io/strimzi/api/kafka/model/AbstractSsLike.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/AbstractSsLike.java]: I miss that! Btw the documentation section "Node and Pod Affinity" mentions Kafka brokers affinity, it's not clear that you can do that with Zookeeper as well even if a `zookeeper-affinity` config param is described in the cluster configmap section.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java (hunk 8 above)
Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java]: I didn't get the need for a field which is private and is not used. You are talking about overriding get/set Metadata which is something you are already doing but this field is not involved at all.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java]: So it can be `CRD_NAME = RESOURCE_NAME` ?

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java (hunk 9 above)
Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java]: K8s takes the stance, and enforces it for CRDs, that extra properties added to the JSON (not defined in the schema) are ignored. You can configure Jackson to do the same, but then when Jackson deserializes a POJO tree it will discard the extra properties, which will then be missing when the POJO tree gets serialized. the additionalProperties (specifically the `@JsonAnyGetter` and `@JsonAnySetter` on the getter and setter) give Jackson somerwhere to put the properties which are not otherwise mapped and means we can round trip the properties we "don't know about" back to JSON.

I don't think this is 100% necessary right now, because we currently only ever deserialize the JSON. I think it will be a problem when:

* we come to support `status` in the CRDs (for which we might need the CRD subresource support in k8s/os, but I've not thought it through in depth yet).
* we want to evolve the CRD schema without incrementing the apiVersion, which is why K8s takes the stance it does.

Basically I think it makes sense if the POJOs support hi-fidelity marshalling to/from JSON right from the start.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java (hunk 10 above)
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: Yeah, thanks! I also intend to change the name at some point because really CRD is an impl detail. The natural name might be `fromAssembly`, but we have one of those already. Maybe `fromResource()`, or `fromDesired()`?

File: api/src/main/java/io/strimzi/api/kafka/model/AbstractSsLike.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/AbstractSsLike.java]: Because in the `KafkaAssembly` (the Java representation of a `kind: Kafka` resource) case it's the superclass of the `Kafka` and `Zookeeper`, which are not assemblies.

But that gives me the idea that `KafkaAssembly` should really be called `KafkaKind`, which is both shorter and more more obvious. Then I would have `KafkaKindOperator`, which resolves a naming conflict with the existing `KafkaAssemblyOperator`. 

Anyway, I thought 'jvm needed to be in there since one of the properties is `jvmOptions`, and 'replicated' because this class also has `replicas`. Really the rubbish name is a reflection of the weak abstraction this class provides: its a base class for things from which we'll generate things which have replicated pods (SS or Deployment) which are running Java.

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java]: Ah, yes. Thanks!

File: api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java (hunk 8 above)
Comment (tombentley) [reply api/src/main/java/io/strimzi/api/kafka/model/KafkaAssembly.java]: Yeah, for whatever reason the field really needs to be there (even though it's unused) for sundrio to generate a build with `withMetadata()`. TBH that might be a bug in sundio, but since I need that method I'm willing to have the redundant field.

File: api/src/test/java/io/strimzi/api/kafka/model/KafkaCrdIT.java [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+@Resources(value = "../examples/install/cluster-operator/07-crd-kafka.yaml", asAdmin = true)
+public class KafkaCrdIT {
+    public static final String NAMESPACE = "kafkacrd-it";
+
+    @ClassRule
+    public static KubeClusterResource cluster = new KubeClusterResource();
+
+    private <T extends CustomResource> void createDelete(Class<T> resourceClass, String resource) {
+        String ssStr = TestUtils.readResource(resourceClass, resource);
+        System.out.println(ssStr);
```
Comment (stanlyDoge) [inline api/src/test/java/io/strimzi/api/kafka/model/KafkaCrdIT.java]: This is just a debug output, right?

**Comment by tombentley** [comment]
Yes, thanks!
